pgxn/neon/walsender_hooks.c

```diff
@@ -0,0 +1,121 @@
+#include "walsender_hooks.h"
+#include "postgres.h"
+#include "fmgr.h"
+#include "access/xlogdefs.h"
+#include "replication/walsender.h"
+#include "access/xlog.h"
+#include "access/xlog_internal.h"
+#include "access/xlogreader.h"
+#include "miscadmin.h"
+#include "utils/wait_event.h"
+
+#include "neon_walreader.h"
+#include "walproposer.h"
+
+static NeonWALReader *wal_reader = NULL;
+extern XLogRecPtr WalSndWaitForWal(XLogRecPtr loc);
+extern bool GetDonorShmem(XLogRecPtr *donor_lsn);
+
+static XLogRecPtr NeonWALReadWaitForWAL(XLogRecPtr loc)
+{
+    while(!NeonWALReaderUpdateDonor(wal_reader))
+        pg_usleep(1000);
+
+    return WalSndWaitForWal(loc);
+}
+
+static int NeonWALPageRead(
+    XLogReaderState *xlogreader,
+    XLogRecPtr targetPagePtr,
+    int reqLen,
+    XLogRecPtr targetRecPtr,
+    char *readBuf)
+{
+    // Wait for flush pointer to advance past our request
+    XLogRecPtr flushptr = NeonWALReadWaitForWAL(targetPagePtr + reqLen);
+    int count;
+    if (flushptr < targetPagePtr + reqLen)
+	return -1;
+
+    // Read at most XLOG_BLCKSZ bytes
+    if (targetPagePtr + XLOG_BLCKSZ <= flushptr)
+	count = XLOG_BLCKSZ;
+    else
+	count = flushptr - targetPagePtr;
+
+    // Sometimes walsender requests non-monotonic sequences of WAL. If that's
+    // the case, we have to reset streaming from remote at the correct position.
+    if(targetPagePtr != NeonWALReaderGetRemLsn(wal_reader))
+    {
+	NeonWALReaderResetRemote(wal_reader);
+    }
+
+    static WaitEventSet *wait_event_set;
+    for(;;)
+    {
+	NeonWALReadResult res = NeonWALRead(
+	    wal_reader,
+	    readBuf,
+	    targetPagePtr,
+	    count,
+	    walprop_pg_get_timeline_id());
+	if(res == NEON_WALREAD_SUCCESS)
+	{
+            // Free event set, since we'll have to recreate it for a new socket later anyway
+	    if (NeonWALReaderEvents(wal_reader) == 0 && wait_event_set != NULL)
+	    {
+		FreeWaitEventSet(wait_event_set);
+		wait_event_set = NULL;
+	    }
+            // Required by xlogreader
+	    xlogreader->seg.ws_tli = NeonWALReaderGetSegment(wal_reader)->ws_tli;
+	    xlogreader->seg.ws_segno = NeonWALReaderGetSegment(wal_reader)->ws_segno;
+	    xlogreader->seg.ws_file = NeonWALReaderGetSegment(wal_reader)->ws_file;
+	    return count;
+	}
+	if(res == NEON_WALREAD_ERROR)
+	{
+	    return -1;
```

Comment:
Out of curiosity I checked what happens without ERROR: XLogReadRecord returns NULL, but that doesn't bother XLogSendLogical so walsender continues to live, apparently creating one more busy loop.

Reply:
Ok, I logged it as ERROR

Reply:
Not sure if I know what you mean about busy loop - it seems that in `XLogSendLogical`, it does 
```
	record = XLogReadRecord(logical_decoding_ctx->reader, &errm);

	/* xlog record was invalid */
	if (errm != NULL)
		elog(ERROR, "%s", errm);
```
so it seems that it also uses elog(ERROR) break out of error state

Reply:
When callback returns XLREAD_FAIL, XLogReadRecord returns NULL with no error message. Which walsender weirdly ignores.

Reply:
Per XLogReadRecord, when callback fails NULL is returned and errormsg is set to null. I haven't checked, maybe this isn't really true and errormsg is not reset, so you could probably communicate it upwards. But even if this is the case, better not to violate the convention since ERROR is fine for us, walsender just dies on it. Note that logical_read_xlog_page essentially does the same in WALReadRaiseError.

Reply:
OK, I think we agree than :laughing: So no further action needed here right?

Reply:
yep